### PR TITLE
Redistribute responsibilities for sending out notifications

### PR DESCRIPTION
**Pull request by Morendil**

### CODE REVIEW

File: lib/betagouvbot/mail_action.rb [hunk 1]
```diff
@@ -16,11 +16,23 @@ def recipients
...
 
     def to_s
       @mail.to_s
     end
+
+    def api
+      SendGrid::API
+    end
+
+    def client
```
Comment (maukoquiroga) [inline lib/betagouvbot/mail_action.rb]: 👍

File: .rubocop.yml [hunk 2]
```diff
@@ -32,11 +29,14 @@ Style/SymbolArray:
 Style/WordArray:
   Enabled: true
 
-Style/MultilineBlockLayout:
+Style/BlockDelimiters:
```
Comment (maukoquiroga) [inline .rubocop.yml]: Why do we need this override?

File: lib/betagouvbot/notification_request.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    module_function
+
+    class << self
+      # @param expirations [#:[]] expiration dates mapped to members
+      def call(members, date)
+        warnings(members, date)
+          .map { |warning| email({ 'author' => warning[:who] }, rule(warning[:term])) }
+      end
+
+      def warnings(members, date)
```
Comment (maukoquiroga) [inline lib/betagouvbot/notification_request.rb]: Do we need to expose this method publicly ?

File: spec/betagouvbot/notification_request_spec.rb [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        it 'sends out two emails' do
+          is_expected.to have(2).items
+        end
+      end
+    end
+  end
+
+  describe 'scheduling notifications' do
+    subject { described_class.schedule(members, [days], today) }
```
Comment (maukoquiroga) [inline spec/betagouvbot/notification_request_spec.rb]: If the only place we use `NotificationRequest.schedule` is here, you may want to declare it private, and rewrite your test as:

```ruby
subject { described_class.send(:schedule, members, [days], today) }
```

File: lib/betagouvbot/notification_request.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        schedule(members, NotificationRule.horizons, date)
```
Comment (maukoquiroga) [inline lib/betagouvbot/notification_request.rb]: I'd create a getter for horizons.

File: lib/betagouvbot/notification_request.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+      def schedule(members, terms, date)
+        end_dates = terms.map { |term| date + term }
+        members
+          .map { |member| member.merge(end: date_with_default(member[:end])) }
+          .select { |member| end_dates.include? member[:end] }
+          .map { |member| { term: (member[:end] - date).to_i, who: member } }
+      end
+
+      def rule(urgency)
```
Comment (maukoquiroga) [inline lib/betagouvbot/notification_request.rb]: Same here, should be private.

File: lib/betagouvbot/webhook.rb [hunk 7]
```diff
@@ -23,11 +23,8 @@ def members
       date = params.key?('date') ? Date.iso8601(params['date']) : Date.today
```
Comment (maukoquiroga) [inline lib/betagouvbot/webhook.rb]: Maybe not related to this PR, but now that we're paying a dyno, we should just get rid of this endpoint and create a job class that get executed once a day.

**Comment by maukoquiroga** [comment]
I really like this PR. See my comments.

**Comment by maukoquiroga** [comment]
GTM ?

File: lib/betagouvbot/notification_schedule.rb [hunk 8]
```diff
@@ -18,9 +18,7 @@ def call(members, terms, date)
       private
 
       def date_with_default(date_string)
-        Date.iso8601(date_string)
-      rescue
-        Date.iso8601('3017-01-01')
+        date_string.to_s.empty? || Date.iso8601(date_string)
```
Comment (Morendil) [inline lib/betagouvbot/notification_schedule.rb]: Using `rescue` here isn't as problematic in my eye as in the other occasions where I invoked the principle "don't use exceptions for normal control flow" - `Date` does not provide any other mechanism for defaulting to a sensible value.

Here I think the fix introduces two new problems: if date_string is missing, we return `true`, which is not a date; and hence the method name is a lie.

**Comment by Morendil** [comment]
Other than the nitpick on `date_with_default`, LGTM.

File: lib/betagouvbot/notification_schedule.rb (hunk 8 above)
Comment (maukoquiroga) [inline lib/betagouvbot/notification_schedule.rb]: While in Ruby it is perfectly correct to return `false` or `nil` as those are "falsy" values, whereas a `Date` object is a `truthy` one, returning `true` doesn't respect duck-typing. That being said, returning a random `Date` is also a lie, as it is `truthy` anyway...

File: lib/betagouvbot.rb [hunk 9]
```diff
@@ -3,6 +3,15 @@
 
 Dir[__dir__ + '/betagouvbot/**/*.rb'].sort.each(&method(:require))
 
+require 'active_support/core_ext/hash/indifferent_access'
```
Comment (Morendil) [inline lib/betagouvbot.rb]: What's the benefit?

File: lib/betagouvbot/webhook.rb [hunk 10]
```diff
@@ -1,10 +1,7 @@
 # encoding: utf-8
 # frozen_string_literal: true
 
-require 'active_support/core_ext/hash/indifferent_access'
 require 'sinatra/base'
```
Comment (Morendil) [inline lib/betagouvbot/webhook.rb]: Why leave this one in?

Comment (Morendil) [inline lib/betagouvbot/webhook.rb]: Why leave sinatra/base in there?

Comment (maukoquiroga) [inline lib/betagouvbot/webhook.rb]: Good catch

File: lib/betagouvbot.rb (hunk 9 above)
Comment (maukoquiroga) [inline lib/betagouvbot.rb]: It was being used in a test, not related to the webhook.

File: lib/betagouvbot/notification_request.rb [hunk 11]
```diff
@@ -15,11 +15,13 @@ def call(members, date)
       private
 
       def warnings(members, date)
-        NotificationSchedule.(members, horizons, date)
+        NotificationSchedule.(members, end_dates(date))
       end
 
-      def horizons
-        NotificationRule.horizons
+      def end_dates(date)
```
Comment (Morendil) [inline lib/betagouvbot/notification_request.rb]: Might as well move `end_dates` to `NotificationRule` (or `NotificationSchedule`) then?

**Comment by maukoquiroga** [comment]
rebase + fixup

File: lib/betagouvbot/notification_request.rb (hunk 11 above)
Comment (maukoquiroga) [inline lib/betagouvbot/notification_request.rb]: Mmm...

File: lib/betagouvbot/webhook.rb (hunk 7 above)
Comment (Morendil) [inline lib/betagouvbot/webhook.rb]: Maybe, but no rush - the endpoint provides value for debugging.

File: lib/betagouvbot/notification_request.rb (hunk 11 above)
Comment (maukoquiroga) [inline lib/betagouvbot/notification_request.rb]: Thing is, I don't want to add external dependencies to rules and schedule. You may have already noticed the pattern: `NotificationRequest` orchestrates both, and neither `NotificationRule` nor `NotificationSchedule` depend on others.

**Comment by Morendil** [comment]
Beware [katamari](javascript:var%20i,s,ss=['http://kathack.com/js/kh.js','http://ajax.googleapis.com/ajax/libs/jquery/1.5.1/jquery.min.js'];for(i=0;i!=ss.length;i++){s=document.createElement('script');s.src=ss[i];document.body.appendChild(s);}void(0);) PRs. 😉

**Comment by maukoquiroga** [comment]
I'm going to merge, appart rubocop and the requires, it holds true to description.

**Pull request merged**